extensions/amp-app-banner/0.1/amp-app-banner.js

```diff
@@ -0,0 +1,177 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {setStyles} from '../../../src/style';
+import {user} from '../../../src/log';
+import {platform} from '../../../src/platform';
+import {viewerFor} from '../../../src/viewer';
+import {CSS} from '../../../build/amp-app-banner-0.1.css';
+import {documentInfoFor} from '../../../src/document-info';
+import {xhrFor} from '../../../src/xhr';
+import {assertHttpsUrl} from '../../../src/url';
+import {isExperimentOn} from '../../../src/experiments';
+
+const TAG = 'amp-app-banner';
+
+class AmpAppBanner extends AMP.BaseElement {
+
+  /** @override */
+  preconnectCallback(onLayout) {
+    if (!this.isExperimentOn_) {
+      user.warn(TAG, `Experiment ${TAG} disabled`);
+      return;
+    }
+
+    if (platform.isIos()) {
+      this.preconnect.url('https://itunes.apple.com', onLayout);
+    } else if (platform.isAndroid()) {
+      this.preconnect.url('https://play.google.com', onLayout);
+    }
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @private @const {boolean} */
+    this.isExperimentOn_ = isExperimentOn(this.getWin(), TAG);
+    if (!this.isExperimentOn_) {
+      user.warn(TAG, `Experiment ${TAG} disabled`);
+      return;
+    }
+
+    if (!platform.isIos() && !platform.isAndroid()) {
+      setStyles(this.element, {
+        'display': 'none',
+      });
+      return;
+    }
+
+    /** @private @const {!Viewer} */
+    const viewer = viewerFor(this.getWin());
+
+    /** @private @const {!Document} */
+    this.doc_ = this.getWin().document;
+
+    /** @private @const {!Xhr} */
+    this.xhr_ = xhrFor(this.getWin());
+
+    /** @const @private {!Element} */
+    this.installLink_ = this.element.querySelector('a[install-link]');
+    user.assert(this.installLink_, '<a install-link> is required inside %s: %s',
+        TAG, this.element);
+
+    /** @const @private {!Element} */
+    this.openLink_ = this.element.querySelector('a[open-link]');
+    user.assert(this.openLink_, '<a open-link> is required inside %s: %s',
+        TAG, this.element);
+
+    if (platform.isIos()) {
+      const meta = this.doc_.head.querySelector(
+          'meta[name=apple-itunes-app]');
+      user.assert(meta,
```

Comment:
Just realized that asserting here might be too strong, since this will force publishers to have BOTH android and ios apps and setup both `meta` and `manifest`. Some might only have one of these so it doesn't make sense to assert here.

@cramforce let me know what do you think?